feeadjuster/feeadjuster.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python3
+import random
+import time
+from pyln.client import Plugin, Millisatoshi, RpcError
+
+
+plugin = Plugin()
+# Our amount and the total amount in each of our channel, indexed by scid
+plugin.adj_balances = {}
+# Cache to avoid loads of calls to getinfo
+plugin.our_node_id = None
+
+
+def get_ratio(our_percentage):
+    """
+    Basic algorithm: the farther we are from the optimal case, the more we
+    bump/lower.
+    """
+    return 50**(0.5 - our_percentage)
+
+
+def maybe_adjust_fees(plugin: Plugin, scids: list):
+    for scid in scids:
+        our = plugin.adj_balances[scid]["our"]
+        total = plugin.adj_balances[scid]["total"]
+        percentage = our / total
+        last_percentage = plugin.adj_balances[scid].get("last_percentage")
+
+        # Only update on substantial balance moves to avoid flooding, and add
+        # some pseudo-randomness to avoid too easy channel balance probing
+        update_treshold = 0.05
+        if not plugin.deactivate_fuzz:
+            update_treshold += random.uniform(-0.015, 0.015)
+
```

Comment:
Timing it by 4 is a huge efficiency decrease, and completely removing the the update between 0.2 and 0.8 kind of defeats the point of using this.. Ideally you would help you to never get to the `[0; 0.2]U[0.8; 1]` zone (which often means for a consequent router an error to forward, with all the consequences).

Reply:
on the other hand its more useful to set custom fees only if the imbalance is high enough. by setting it when you are at 40/60 you gain close to nothing but reveal you balance...

Reply:
I think we should consider hysteresis as well.  For example, have a low/med/high feerate.  Then:

    if feerate is "low":
        if balance < 0.25:
            feerate = "high"
        elif balance < 0.5:
            feerate = "med"
    elif feerate is "med":
        if balance < 0.25:
            feerate = "high"
        elif balance > 0.75:
            feerate = "low"
    else: # feerate is "high":
        if balance > 0.75:
            feerate = "low"
        elif balance > 0.5:
            feerate = "med"

This gives better stability in case of noise in the balance signal.

Though as @darosior points out, it would be nice to have finer divisions so you never even enter the extreme zones.

Reply:
> I think we should consider hysteresis as well

Don't you think the update threshold is hysterisis ?

> Though as @darosior points out, it would be nice to have finer divisions so you never even enter the extreme zones.

Yeah, so i think that instead of having fine-grained thresholds for performing the update, it's better to have thresholds for not performing this update (which is similar, but the logic is cleaner).
Also, this threshold is fuzzed in this PR.

Reply:
>Don't you think the update threshold is hysterisis ?

My python-fu is awful, so I have no real idea.

More honestly, I have not actually read any of the code LOL.

Reply:
On the other hand, IMO *any* feeadjuster is better than none, feel free to ignore me here, I have no idea about the issues a consequential router has.

Reply:
I might add that run in current form, it reveals the channel balance with an accuracy of >90% when monitored over a certain period of time:
 - find out base fees by taking median of all channels... repeat over time
 - apply reverse calculation on each channel
 - ignore the 1.5% fuzz ;)

Don't get me wrong, I really like it, hence I use it (with a threshhold). Thus we need to improve.

Reply:
> it reveals the channel balance with an accuracy of >90% when monitored over a certain period of time:
> find out base fees by taking median of all channels... repeat over time
> apply reverse calculation on each channel

Did you do that ? Didn't the (fuzzed) 5% threshold messed your computation ?

Reply:
Nope, didn't do it tbh. But wouldn't that simply reduce accuracy by another 5%?

Reply:
>  On the other hand, the existence of the threshold in the first place is the hystersis, I think....... hmmm.

Yeah, it's just that i fuzz the hysteresis (if that makes any sense). 

> I would go for a randomized-hourly update myself rather than on a forward notification but shrug --- I have not run a "consequential" router, so my thoughts may be bunk here

That decision was not based on any needs but to:
- Avoid scheduled (thus predictable?) updates
- Flatten load (don't broadcast 2 * n_channels update each hour)
- Be more accurate

> Another is that I would use absolute bin sizes instead of relative percent-based ones

Hmm in this case i guess their size should increase exponentially (as bitcoind fee estimation buckets)

> I have no idea about the issues a consequential router has.

This is also for us small-router operators! :p
I look forward to run (a correct version of) this on my node too and i hear your concerns regarding balance de-obfuscation.

However, i fail to see how a bucket system would be more private than the current threshold (which is basically, have buckets of size 5% (fuzzed)).

In fact, i fail to see how any trick we could add would make this more private by not losing efficiency: this **is** about sharing our balance.
I think we should instead concentrate on making it not directly gamable by probing.

Reply:
> However, i fail to see how a bucket system would be more private than the current threshold (which is basically, have buckets of size 5% (fuzzed)).

When you move out of the current bucket, you then set the fee accurately according to the exact balance percent.  Thus, by reversing your computation, a third party knows thee exact balance percent at the moment you change the fee.

With a strict bucket system, once you move from one bucket to another, you set the fee according to the *center* of the bucket you entered, not the actual current exact balance percent.  This leaks less accurate information when you move out of your current bucket.

Now, if your exact balance moves, but still within the current bucket, you will not change the fee.  This leads to an inefficiency.  Thus, you would suffer inefficiency either way, especially if there is a large amount of traffic through your channels, when the balance moves a small amount.  But with a strict bucket system, you do not leak your exact balance at the time you change fees, only within the bucket size, which are the error bars that a third party can use to monitor you.

This is probably just a minor effect however.